doc: add a section about coverity scans (#58749)
zephyrproject-rtos/zephyr: doc/contribute/guidelines.rst

A more complete alternative to this is using check_compliance.py script from
ci-tools repo.

Static Code Analysis
********************

Coverity Scan is a free service for static code analysis of Open Source
projects. It is based on Coverity's commercial product and is able to analyze
C, C++ and Java code.

Coverity's static code analysis doesn't run the code. Instead of that it uses
abstract interpretation to gain information about the code's control flow and
data flow. It's able to follow all possible code paths that a program may take.
For example the analyzer understands that malloc() returns a memory that must
be freed with free() later. It follows all branches and function calls to see
if all possible combinations free the memory. The analyzer is able to detect
all sorts of issues like resource leaks (memory, file descriptors), NULL
dereferencing, use after free, unchecked return values, dead code, buffer
overflows, integer overflows, uninitialized variables, and many more.

> The results are available on the `Coverity Scan <https://scan.coverity.com/>`_

Reviewer: The project URL is public so prob better to link to it directly?

```suggestion
The results are available on the `Coverity Scan <https://scan.coverity.com/projects/zephyr>`_
```

website. In order to access the results you have to create an account yourself.

> Then go to Projects using Scan and add yourself to the Zephyr project. New

Reviewer: can be reworded if link above is changed to a direct link to the Zephyr scan. 

```suggestion
From the Zephyr project page, you may select "Add me to project" to be added to the project. New
```

members must be approved by an admin.

The Zephyr codebase is scanned frequently (Weekly) and results of the scan are
processed and issues are created automatically as bugs. Bugs are assigned
automatically to the maintainers of the areas affected.

Known limitations
=================

Placeholder for list of known issues and limitations of the scan service.

Workflow
========

> If the problem is listed under Known limitations then please set the

Reviewer 2: what are "Known limitations" in Zephyr?

Reviewer 2: shouldn't we say "If, after analyzing the Coverity report, you come to the conclusion that it is a false positive, ..." ?

Author: > what are "Known limitations" in Zephyr?

Is is not about zephyr, it is limiation of tool itself. I added a section above which will be populated. For example, coverity, it seems, has an assue with zephyr ASSERTs which might result in lots of false positives. We do not check for errors and rely on ASSERTs,  and the tool thinks it is a problem.

Author: > what are "Known limitations" in Zephyr?

Is is not about zephyr, it is limiation of tool itself. I added a section above which will be populated. For example, coverity, it seems, has an assue with zephyr ASSERTs which might result in lots of false positives. We do not check for errors and rely on ASSERTs,  and the tool thinks it is a problem.

classification to either "False positive" or "Intentional", the action to
"Ignore", owner to your own account and add a comment why the issue is
considered false positive or intentional.

Update the related issue in the Zephyr project with the details and only close
the issue after completing the step above on scan service website. Any issues

> closed without a fix or without ignore the entry in the scan service will be

Reviewer 2: ```suggestion
closed without a fix or without ignoring the entry in the scan service will be
```
